### PR TITLE
feat(v2): Allow plugins to consume webpack stats

### PR DESCRIPTION
**Pull request by RDIL**
Signed-off-by: Reece Dunham <me@rdil.rocks>

## Motivation

I need to use one of these stats in my own plugin, so it was `patch-package`, or do this, which gives plugins a LOT more options for post-processing and other capabilities.  

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

My test plugin worked against the branch.

## Related PRs

n/a

### CODE REVIEW

File: packages/docusaurus/src/commands/start.ts [hunk 1]
```diff
@@ -103,7 +103,7 @@ export default async function start(
     cwd: siteDir,
     ignoreInitial: true,
     usePolling: !!cliOptions.poll,
-    interval: Number.isInteger(cliOptions.poll)
+    interval: Number.isInteger((cliOptions.poll as unknown) as never)
```
Comment (slorber) [inline packages/docusaurus/src/commands/start.ts:106]: what was the problem?

**Review by slorber — COMMENTED** [review]
LGTM, thanks 👍 

Was wondering why the start.ts type is updated?

File: packages/docusaurus/src/webpack/utils.ts [hunk 2]
```diff
@@ -198,22 +198,24 @@ function filterWarnings(
...
-        stats.toJson('errors-only').errors.forEach((e) => {
+        allStats.errors.forEach((e) => {
           console.error(e);
         });
         reject(new Error('Failed to compile with errors.'));
       }
       if (stats?.hasWarnings()) {
         // Custom filtering warnings (see https://github.com/webpack/webpack/issues/7841).
-        let {warnings} = stats.toJson('errors-warnings');
+        let warnings = [...allStats.warnings];
```
Comment (slorber) [inline packages/docusaurus/src/webpack/utils.ts:218]: are there cases where the warning array is not defined?

Maybe we could replace the condition with `if (allStats.warnings?.length > 0)`, and TS would understand better that the array is there or something?

File: packages/docusaurus/src/commands/start.ts (hunk 1 above)
Comment (RDIL) [reply packages/docusaurus/src/commands/start.ts:106]: I thought I was going to need to update this for this PR, and my IDE was complaining about it.  I'll revert if you want.

File: packages/docusaurus/src/webpack/utils.ts (hunk 2 above)
Comment (RDIL) [reply packages/docusaurus/src/webpack/utils.ts:218]: This is there so the original `stats.warnings` doesn't get modified.

File: packages/docusaurus/src/commands/start.ts (hunk 1 above)
Comment (slorber) [reply packages/docusaurus/src/commands/start.ts:106]: As it's not strictly related to the PR and I don't understand the intent, I'd prefer removing this line and discuss it as a separate issue/PR if needed

Comment (RDIL) [reply packages/docusaurus/src/commands/start.ts:106]: Sure.

File: packages/docusaurus/src/webpack/utils.ts (hunk 2 above)
Comment (slorber) [reply packages/docusaurus/src/webpack/utils.ts:218]: I see thanks.
It shouldn't be, but won't harm to keep this code

File: packages/docusaurus/src/commands/start.ts (hunk 1 above)
Comment (RDIL) [reply packages/docusaurus/src/commands/start.ts]: Done.

**Comment by slorber** [comment]
thanks 👍

**Comment by RDIL** [comment]
No, thank you!  My plugin is now a little closer to completion 😄

**Comment by slorber** [comment]
great to know 👍 add it to the community section when ready

**Pull request merged**